### DISCUSSION TITLE
When using vSphere 6.7, not able pass kickstart file try both disk and http after setting firmware to BIOS

### DISCUSSION THREAD

**Discussion by TouqeerHussain** [vmware/packer-examples-for-vsphere#168] [Help]
Hi There, can you please help me to identified what i am missing , somehow the build start bit not able to pass the kickstart file tried both methods

ENV is 6.7

Packer Inspect: HCL2 mode

> input-variables:

var.<sensitive>_key: "XXXXX"
var.<sensitive>_username: "<sensitive>"
var.build_key: "XXXXX"
var.build_password: "<sensitive>"
var.build_password_encrypted: "<sensitive>"
var.build_username: "<sensitive>"
var.common_content_library_destroy: "true"
var.common_content_library_name: "lib01"
var.common_content_library_ovf: "true"
var.common_content_library_skip_export: "false"
var.common_data_source: "disk"
var.common_http_ip: "null"
var.common_http_port_max: "8080"
var.common_http_port_min: "8080"
var.common_ip_wait_timeout: "20m"
var.common_iso_datastore: "datastore3"
var.common_remove_cdrom: "true"
var.common_shutdown_timeout: "15m"
var.common_template_conversion: "false"
var.common_tools_upgrade_policy: "true"
var.common_vm_version: "13"
var.communicator_port: "22"
var.communicator_proxy_host: "null"
var.communicator_proxy_password: "null"
var.communicator_proxy_port: "null"
var.communicator_proxy_username: "null"
var.communicator_timeout: "30m"
var.iso_checksum_type: "sha256"
var.iso_checksum_value: "19d653ce2f04f202e79773a0cbeda82070e7527557e814ebbce658773fbe8191"
var.iso_file: "rhel-server-7.9-x86_64-dvd.iso"
var.iso_path: "iso/linux/rhel"
var.rhsm_password: "<sensitive>"
var.rhsm_username: "<sensitive>"
var.vm_boot_order: "disk,cdrom"
var.vm_boot_wait: "2s"
var.vm_cdrom_type: "sata"
var.vm_cpu_cores: "1"
var.vm_cpu_hot_add: "false"
var.vm_cpu_sockets: "2"
var.vm_disk_controller_type: "[\n  \"pvscsi\",\n]"
var.vm_disk_size: "40960"
var.vm_disk_thin_provisioned: "true"
var.vm_firmware: "bios"
var.vm_guest_os_family: "linux"
var.vm_guest_os_keyboard: "GB"
var.vm_guest_os_language: "en_GB"
var.vm_guest_os_name: "rhel"
var.vm_guest_os_timezone: "UTC"
var.vm_guest_os_type: "rhel7_64Guest"
var.vm_guest_os_version: "7"
var.vm_mem_hot_add: "false"
var.vm_mem_size: "2048"
var.vm_network_card: "vmxnet3"
var.vsphere_cluster: "LONDGEN01"
var.vsphere_datacenter: "LON"
var.vsphere_datastore: "datastore3"
var.vsphere_endpoint: "XXXXX"
var.vsphere_folder: "templates"
var.vsphere_insecure_connection: "true"
var.vsphere_network: "VMNetwork"
var.vsphere_password: "<sensitive>"
var.vsphere_username: "<sensitive>"

> local-variables:

local.build_by: "Built by: HashiCorp Packer 1.7.10"
local.build_date: "2022-03-08 13:37 UTC"
local.build_version: "22.03"
local.data_source_command: "inst.ks=cdrom:/ks.cfg"
local.data_source_content: "{\n  \"/ks.cfg\" = \"# THE SOFTWARE IS PROVIDED \\\"AS IS\\\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE\\n# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\\n# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR\\n# OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\\n\\n# Red Hat Enterprise Linux Server 7\\n\\n### Installs from the first attached CD-ROM/DVD on the system.\\ncdrom\\n\\n### Performs the kickstart installation in text mode. \\n### By default, kickstart installations are performed in graphical mode.\\ntext\\n\\n### Accepts the End User License Agreement.\\neula --agreed\\n\\n### Sets the language to use during installation and the default language to use on the installed system.\\nlang en_GB\\n\\n### Sets the default keyboard type for the system.\\nkeyboard GB\\n\\n### Configure network information for target system and activate network devices in the installer environment (optional)\\n### --onboot\\t  enable device at a boot time\\n### --device\\t  device to be activated and / or configured with the network command\\n### --bootproto\\t  method to obtain networking configuration for device (default dhcp)\\n### --noipv6\\t  disable IPv6 on this device\\n###\\n### network  --bootproto=static --ip=172.16.11.200 --netmask=255.255.255.0 --gateway=172.16.11.200 --nameserver=172.16.11.4 --hostname centos-linux-8\\nnetwork --bootproto=dhcp\\n\\n### Lock the root account.\\nrootpw --lock\\n\\n### The selected profile will restrict root login.\\n### Add a user that can login and escalate privileges.\\nuser --name=<sensitive> --iscrypted --password=<sensitive> --groups=wheel\\n\\n### Configure firewall settings for the system.\\n### --enabled\\treject incoming connections that are not in response to outbound requests\\n### --ssh\\t\\tallow sshd service through the firewall\\nfirewall --enabled --ssh\\n\\n### Sets up the authentication options for the system.\\n### Sets sha512 to hash passwords. Passwords are shadowed.\\nauth --passalgo=sha512 --useshadow\\n\\n### Sets the state of SELinux on the installed system.\\n### Defaults to enforcing.\\nselinux --enforcing\\n\\n### Sets the system time zone.\\ntimezone UTC\\n\\n### Sets how the boot loader should be installed.\\nbootloader --location=mbr\\n\\n### Initialize any invalid partition tables found on disks.\\nzerombr\\n\\n### Removes partitions from the system, prior to creation of new partitions. \\n### By default, no partitions are removed.\\n### --linux\\terases all Linux partitions.\\n### --initlabel Initializes a disk (or disks) by creating a default disk label for all disks in their respective architecture.\\nclearpart --all --initlabel\\n\\n### Modify partition sizes for the virtual machine hardware.\\n### Create primary system partitions.\\npart /boot --fstype xfs --size=1024 --label=BOOTFS\\npart /boot/efi --fstype vfat --size=1024 --label=EFIFS\\npart pv.01 --size=100 --grow\\n\\n### Create a logical volume management (LVM) group.\\nvolgroup sysvg --pesize=4096 pv.01\\n\\n### Modify logical volume sizes for the virtual machine hardware.\\n### Create logical volumes.\\nlogvol swap --fstype swap --name=lv_swap --vgname=sysvg --size=1024 --label=SWAPFS\\nlogvol / --fstype xfs --name=lv_root --vgname=sysvg --size=12288 --label=ROOTFS\\nlogvol /home --fstype xfs --name=lv_home --vgname=sysvg --size=4096 --label=HOMEFS --fsoptions=\\\"nodev,nosuid\\\"\\nlogvol /opt --fstype xfs --name=lv_opt --vgname=sysvg --size=2048 --label=OPTFS --fsoptions=\\\"nodev\\\"\\nlogvol /tmp --fstype xfs --name=lv_tmp --vgname=sysvg --size=4096 --label=TMPFS --fsoptions=\\\"nodev,noexec,nosuid\\\"\\nlogvol /var --fstype xfs --name=lv_var --vgname=sysvg --size=4096 --label=VARFS --fsoptions=\\\"nodev\\\"\\nlogvol /var/log --fstype xfs --name=lv_log --vgname=sysvg --size=4096 --label=LOGFS --fsoptions=\\\"nodev,noexec,nosuid\\\"\\nlogvol /var/log/audit --fstype xfs --name=lv_audit --vgname=sysvg --size=4096 --label=AUDITFS --fsoptions=\\\"nodev,noexec,nosuid\\\"\\n\\n### Modifies the default set of services that will run under the default runlevel.\\nservices --enabled=NetworkManager,sshd\\n\\n### Do not configure X on the installed system.\\nskipx\\n\\n### Packages selection.\\n%packages --ignoremissing --excludedocs\\n@core\\n-iwl*firmware\\n%end\\n\\n### Post-installation commands.\\n%post\\n/usr/sbin/subscription-manager register --username <sensitive> --password <sensitive> --autosubscribe --force\\n/usr/sbin/subscription-manager repos --enable \\\"rhel--optional-rpms\\\" --enable \\\"rhel--extras-rpms\\\" --enable \\\"rhel-ha-for-rhel-*-server-rpms\\\"\\nyum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm\\nyum makecache\\nyum install -y sudo open-vm-tools perl <sensitive>\\necho \\\"<sensitive> ALL=(ALL) NOPASSWD: ALL\\\" >> /etc/sudoers.d/<sensitive>\\nsed -i \\\"s/^.*requiretty/#Defaults requiretty/\\\" /etc/sudoers\\n%end\\n\\n### Reboot after the installation is complete.\\n### --eject attempt to eject the media before rebooting.\\nreboot --eject\"\n}"
local.manifest_date: "2022-03-08 13:37:58"
local.manifest_path: "/opt/ebrd-packer-build/manifests/"

> builds:

  > <unnamed build 0>:

    sources:

      vsphere-iso.linux-rhel

    provisioners:

      <sensitive>

    post-processors:

      0:
        manifest

**Comment by TouqeerHussain** [comment]
@tenthirtyam can you please help/guide with above issue?

**Comment by TouqeerHussain** [comment]
vsphere-iso.linux-rhel: output will be in this color.

==> vsphere-iso.linux-rhel: Creating CD disk...
    vsphere-iso.linux-rhel: Warning: creating filesystem with Joliet extensions but without Rock Ridge
    vsphere-iso.linux-rhel: extensions. It is highly recommended to add Rock Ridge.
    vsphere-iso.linux-rhel: I: -input-charset not specified, using utf-8 (detected in locale settings)
    vsphere-iso.linux-rhel: Total translation table size: 0
    vsphere-iso.linux-rhel: Total rockridge attributes bytes: 0
    vsphere-iso.linux-rhel: Total directory bytes: 0
    vsphere-iso.linux-rhel: Path table size(bytes): 10
    vsphere-iso.linux-rhel: Max brk space used 0
    vsphere-iso.linux-rhel: 183 extents written (0 MB)
    vsphere-iso.linux-rhel: Done copying paths from CD_dirs
==> vsphere-iso.linux-rhel: Uploading packer438194771.iso to packer_cache/packer438194771.iso
==> vsphere-iso.linux-rhel: Creating VM...
==> vsphere-iso.linux-rhel: Customizing hardware...
==> vsphere-iso.linux-rhel: Adding SATA controller...
==> vsphere-iso.linux-rhel: Mounting ISO images...
==> vsphere-iso.linux-rhel: Adding configuration parameters...
==> vsphere-iso.linux-rhel: Set boot order...
==> vsphere-iso.linux-rhel: Power on VM...
==> vsphere-iso.linux-rhel: Waiting 2s for boot...
==> vsphere-iso.linux-rhel: Typing boot command...
==> vsphere-iso.linux-rhel: Waiting for IP...
Cancelling build after receiving interrupt
==> vsphere-iso.linux-rhel: Power off VM...
==> vsphere-iso.linux-rhel: Destroying VM...
==> vsphere-iso.linux-rhel: Deleting cd_files image from remote datastore ...
Build 'vsphere-iso.linux-rhel' finished after 3 minutes 28 seconds.

**Comment by tenthirtyam** [comment]
The project is configured to **only** use EFI / EFI Secure boot. You have set `var.vm_firmware = "bios"`.

As a result, the boot command and kickstart cfg would need to be changed to support BIOS. 

Please note:
- There are no plans to amend this project to enable both EFI and BIOS. It recently moved only using EFI.
- There are no current plans to relax the requirements to work with pre-vSphere 7.0 U2 due to some content library issues on prior versions.

Ryan

**Comment by TouqeerHussain** [comment]
@tenthirtyam Thank you you are lifesaver. can you please give me another favor , do you have any sample which can i use to amend boot command

**Reply by tenthirtyam** [reply]
Not offhand, no complete sample.

 You can search the project commit history and pull requests to research the changes to EFI but much has changed.

Each OS has its own requirements for BIOs vs EFI boot commands.

For example, RHEL7 boot cmd might look like `<tab>","text ks=http://fqdn:port/ks.cfg","<enter><wait>`, but the ks.cfg would be different for the disk layout like 

```
bootloader --location=mbr --append="rhgb quiet crashkernel=auto"
zerombr
autopart
clearpart --all --initlabel
```
